### DISCUSSION TITLE
Getting GEOS cubed-sphere History output into uxarray easily

### DISCUSSION THREAD

**Discussion by mathomp4** [GEOS-ESM/MAPL#2844] [History]
Back in April I attended the ISS Conference in Boulder and one of the subjects I was most intrigued by was on [UXarray](https://github.com/UXARRAY/uxarray) by @philipc2 and @erogluorhan ([tutorial](https://youtu.be/S1NR3QrWnaY?si=OSEQyuaQuHvViurO&t=18491) and [talk](https://youtu.be/TErVJz7rtqI?si=HAwWfG-a3muKDAMc&t=2445)).

Now, our cubed-sphere format is in a way an unstructured grid. Well, more unstructured than boring old lat-lon, enough that most tools for lat-lon viewing can't understand it. (The one major exception other than home-grown tools is [Panoply](https://www.giss.nasa.gov/tools/panoply/) which @rschmunk was nice enough to work with us to get it to recognize our format.)

UXarray already sort of has an "easy" mode for MPAS, SCRIP, and ESMF grid (and other) formats. I believe internally it can recognize these formats and then converts to [UGRID format](https://ugrid-conventions.github.io/ugrid-conventions/). 

So, my thought is in order to get GEOS CS support in UXarray, our main challenge is figuring out how to get GEOS CS into UGRID format (a la https://github.com/UXARRAY/uxarray/issues/773 for ICON).

Good thing: we have people like @bena-nasa who knows SCRIP and ESMF pretty well and so the fact that there is SCRIP-to-UGRID conversion somewhere in there, might mean it can be sort of "reverse engineered" to a GEOS-to-UGRID converter.

**Comment by philipc2** [comment]
Hi @mathomp4 

I'm glad you found the talks during ISS interesting! Always happy to hear from the community. 

> Good thing: we have people like @bena-nasa who knows SCRIP and ESMF pretty well and so the fact that there is SCRIP-to-UGRID conversion somewhere in there, might mean it can be sort of "reverse engineered" to a GEOS-to-UGRID converter.

Yeah, ESMF grids are very similar to UGRID, with minor modifications needed to get them represent in the UGRID conventions. 

If you have a sample grid that I could take a look at, I should be able to get a rough idea on how we can move forward with implementing a GEOS CS grid reader.

**Reply by mathomp4** [reply]
@philipc2 Okay. Let me see if I can whip up a small exemplar. Is netCDF4 output okay?

**Reply by philipc2** [reply]
Yep! NetCDF is perfect.

**Reply by mathomp4** [reply]
Please find attached a gzipped netCDF file (only so it can be uploaded):

[test-c12.native.nc4.gz](https://github.com/GEOS-ESM/MAPL/files/15422095/test-c12.native.nc4.gz)

This is a *very* coarse file, c12 (so 12x12 on each face of the cube), but that makes it pretty easy to read.

Let us know if something else might be needed.

**Reply by philipc2** [reply]
Thank you! Having a coarse grid is great, especially if we chose to include it in our testing suite.

I'll take a look into it and get back soon!

**Reply by philipc2** [reply]
Was able to generate the following mesh! 

<img width="340" alt="image" src="https://github.com/GEOS-ESM/MAPL/assets/67855069/fbca7aea-6fd8-49ca-9390-5ca9c3eb6d22">

<img width="991" alt="image" src="https://github.com/GEOS-ESM/MAPL/assets/67855069/3fc22aa8-c322-405d-b87b-8484a5577a84">

Also plotting "T". There were multiple `nan` values, which I set to zero.

<img width="996" alt="image" src="https://github.com/GEOS-ESM/MAPL/assets/67855069/c2a781dc-f781-4623-a32a-8b6142bdbaa5">

**Reply by philipc2** [reply]
Here's a minimal example. The code is by no means clean or optimized, but it works.

```Python

import uxarray as ux
import xarray as xr
import numpy as np

xrds = xr.open_dataset("/Users/philipc/Downloads/test-c12.native.nc4")

corner_lons = xrds['corner_lons'].values
corner_lats = xrds['corner_lats'].values
node_lon = corner_lons.flatten()
node_lat = corner_lats.flatten()

face_node_connectivity = []

nf, nx, ny = corner_lons.shape

w = nx

for k in range(0, nf):
    for i in range(nx-1):
        for j in range(ny-1):
            tl = (i * w + j) + (k * w * w)
            tr = (i * w + (j + 1)) + (k * w * w)
            bl = ((i + 1) * w + j) + (k * w * w)
            br = ((i + 1) * w + (j + 1)) + (k * w * w)
            
            face_node_connectivity.append([br, bl, tl, tr])

face_node_connectivity = np.asarray(face_node_connectivity)  

uxgrid = Grid.from_topology(node_lon, node_lat, face_node_connectivity)

uxgrid.plot()


```

**Reply by mathomp4** [reply]
Oooh. Very cool! And, yes, the T you plotted is for 1000 mb, so undef values where terrain makes that impossible. If you went up in the atmosphere, you (should) get defined values everywhere. 

I'll give this a try next week with some higher res data. And we have stretched grid output as well which I think will look very cool

I'll also mention @JulesKouatchou  here just because I'll probably be coming to him with xarray questions 🙂

**Reply by philipc2** [reply]
Going to be continuing this in https://github.com/UXARRAY/uxarray/pull/802

**Reply by philipc2** [reply]
<img width="1114" alt="image" src="https://github.com/GEOS-ESM/MAPL/assets/67855069/f7dc45c8-2789-4a41-9927-71a790418a2a">

**Reply by philipc2** [reply]
<img width="735" alt="image" src="https://github.com/GEOS-ESM/MAPL/assets/67855069/0a069b72-3acb-43f1-a459-a781d8c17049">

**Reply by mathomp4** [reply]
Wow. We do not have much topography at C12! All the mountain ranges are just blobs. 😆 

@philipc2 Maybe once you think you have something good, I can find much higher resolution CS output so we can try for a prettier picture. 

For example, @wmputman does runs at C2160 which is a stretched grid cubed-sphere of ~2km compared to the juuuuuuuuust slightly coarser 750-km resolution of C12. 😃

**Reply by philipc2** [reply]
I'd love to try out a higher resolution example! It'll give me a better idea on the expected performance too.

**Reply by mathomp4** [reply]
Sounds good. It'll probably be next week before I can get something. I have to remind myself of how to transfer/link files to outside of NASA. It seems to change each time I need to do it!

**Reply by philipc2** [reply]
No worries! Thanks!

**Reply by mathomp4** [reply]
@philipc2 Okay. There were some very-high-res files already out there, just needed to be pointed to them. I present [GEOS DYAMOND2](https://gmao.gsfc.nasa.gov/global_mesoscale/dyamond_phaseII/)!

These are truly excitingly large files. For example, this:

https://portal.nccs.nasa.gov/datashare/G5NR/DYAMONDv2/01KM/DYAMONDv2_c5760_L181/inst_15mn_2d_asm_Mx/202002/DYAMONDv2_c5760_L181.inst_15mn_2d_asm_Mx.20200202_1200z.nc4

is a ~21 GB file with 2D diagnostics on our C5760 grid (aka 5760x5760x6).

If that is too big, we have smaller (in resolution) output as well (at C2880 and C1440).

**Reply by mathomp4** [reply]
Note: these aren't stretched grid. I'm still looking at trying to get you something from that, but I've encountered a few runtime...oddities in trying to generate some on my own.

**Reply by philipc2** [reply]
Sweet!! I've worked with some data from the DAMOND initiative before (a 3.75km MPAS mesh), excited to try this one out!

**Reply by philipc2** [reply]
Woah, this is definitely the largest grid I've ever worked with. 

```
<uxarray.Grid>
Original Grid Type: GEOS-CS
Grid Dimensions:
  * n_node: 199134726
  * n_face: 199065600
  * n_max_face_nodes: 4
  * n_nodes_per_face: (199065600,)
Grid Coordinates (Spherical):
  * node_lon: (199134726,)
  * node_lat: (199134726,)
  * face_lon: (199065600,)
  * face_lat: (199065600,)
Grid Coordinates (Cartesian):
Grid Connectivity Variables:
  * face_node_connectivity: (199065600, 4)
```

Rasterizing the cell-centers is extremely quick (roughly a few seconds) for generating a global plot

![image](https://github.com/GEOS-ESM/MAPL/assets/67855069/bb6290b2-4e0a-4f26-8bde-6b8c865d05da)

However, zooming in, much of the detail is lost.

![image](https://github.com/GEOS-ESM/MAPL/assets/67855069/d72887f1-2178-433b-8050-ad02efb792bb)


Using a larger pixel ratio helps, at the cost of execution time.

![image](https://github.com/GEOS-ESM/MAPL/assets/67855069/c9dfaaef-853c-447f-8e44-030ec224b845)


Much better local fidelity

![image](https://github.com/GEOS-ESM/MAPL/assets/67855069/33b8c921-f2e8-42fa-bbcf-ea820e7cdd40)

We also support polygon plotting with UXarray, but it is very impractical for the scale of this dataset. Going to look into taking a regional subset and doing a polygon plot on that region.

**Reply by mathomp4** [reply]
Very cool! And, yes, big. Many tools seem to choke on files this size. I suppose if there could be some...dynamic way of doing things? Raster at global scale but somehow transition to polygons at a finer scale? You are the experts!

Luckily, this is an extreme case. Our current operational run is at C720 and maybe someday we will move to C1440. But even at C720 you are near the size of pixels in most PNGs one produces.

**Reply by philipc2** [reply]
> I suppose if there could be some...dynamic way of doing things? Raster at global scale but somehow transition to polygons at a finer scale? You are the experts!

Yeah this is something that Datashader & HoloViews enables us to do with our UXarray visualization API. It does however require some processing to get the grid represented as polygons. 

This is what a 3.75km MPAS mesh looks like ([from our Cookbook](https://projectpythia.org/unstructured-grid-viz-cookbook/notebooks/03-uxarray-vis/06-performance.html))

![image](https://github.com/GEOS-ESM/MAPL/assets/67855069/ce2b8179-6add-4ec9-b3db-91dc6874d037)

![image](https://github.com/GEOS-ESM/MAPL/assets/67855069/14eb9e2a-f9af-48f5-8e40-fea5bcad86c1)

This grid is about 42 million faces and 84 million nodes, which is smaller than the C1440 grid. Takes about 93s to process the grid into polygons (a one-time operation), and then subsequent visualizations can be rendered in a few seconds (including dynamic rasterization when zooming / panning).

**Comment by rschmunk** [comment]
In something of a coincidence, I have been working on getting Panoply to plot UGRID netCDF data. Still at the incremental stage where, okay, I need an another example file in order to move on to the next improvement. But getting there.

**Reply by mathomp4** [reply]
It's all coming together!

<a href="https://imgflip.com/i/8rfmhe"><img src="https://i.imgflip.com/8rfmhe.jpg" title="made at imgflip.com"/></a><div><a href="https://imgflip.com/memegenerator">from Imgflip Meme Generator</a></div>
